File: agent/scheduler/system_jobs.py
from __future__ import annotations

from .models import SchedulerJob, SchedulerPayload, SchedulerSchedule

SYSTEM_JOB_IDS = {
    "memory-maintenance",
    "runtime-maintenance",
    "team-stale-recovery",
    "token-ledger-maintenance",
    "watchlist-check",
}


def default_system_jobs(*, now: int) -> list[SchedulerJob]:
    return [
        SchedulerJob.create(
            job_id="memory-maintenance",
            name="Memory maintenance",
            schedule=SchedulerSchedule(kind="cron", expr="17 3 * * *", tz="Asia/Shanghai"),
            payload=SchedulerPayload(
                kind="system_event",
                message="memory-maintenance",
                meta={"system_event": "memory-maintenance"},
            ),
            protected=True,
            purpose="检查热 history 与长期记忆维护状态，后续可挂载自动压缩策略。",
            now=now,
        ),
        SchedulerJob.create(
            job_id="runtime-maintenance",
            name="Runtime event maintenance",
            schedule=SchedulerSchedule(kind="cron", expr="37 3 * * *", tz="Asia/Shanghai"),
            payload=SchedulerPayload(
                kind="system_event",
                message="runtime-maintenance",
                meta={"system_event": "runtime-maintenance"},
            ),
            protected=True,
            purpose="检查 runtime/events.jsonl 冷记录规模，后续可挂载轮转策略。",
            now=now,
        ),
        SchedulerJob.create(
            job_id="team-stale-recovery",
            name="Team stale recovery",
            schedule=SchedulerSchedule(kind="every", every_ms=60 * 60 * 1000),
            payload=SchedulerPayload(
                kind="system_event",
                message="team-stale-recovery",
                meta={"system_event": "team-stale-recovery"},
            ),
            protected=True,
            purpose="检查持久队友 stale working 状态，避免下次唤醒前状态不可见。",
            now=now,
        ),
        SchedulerJob.create(
            job_id="token-ledger-maintenance",
            name="Token ledger maintenance",
            schedule=SchedulerSchedule(kind="cron", expr="47 3 * * *", tz="Asia/Shanghai"),
            payload=SchedulerPayload(
                kind="system_event",
                message="token-ledger-maintenance",
                meta={"system_event": "token-ledger-maintenance"},
            ),
            protected=True,
            purpose="检查 tokens.jsonl 账本规模与缓存统计，后续可挂载归档策略。",
            now=now,
        ),
        SchedulerJob.create(
            job_id="watchlist-check",
            name="Watchlist heartbeat",
            schedule=SchedulerSchedule(kind="every", every_ms=6 * 60 * 60 * 1000),
            payload=SchedulerPayload(
                kind="system_event",
                message="watchlist-check",
                meta={"system_event": "watchlist-check"},
            ),
            protected=True,
            purpose="周期检查 memory/watchlist.md，必要时触发本地主动 turn。",
            now=now,
        ),
    ]


def is_system_job(job_id: str) -> bool:
    return str(job_id or "") in SYSTEM_JOB_IDS

File: packages/core/src/scheduler/system-jobs.ts
import { SchedulerJob, SchedulerPayload, SchedulerSchedule } from './models'

export const SYSTEM_JOB_IDS = new Set([
  'memory-maintenance',
  'runtime-maintenance',
  'team-stale-recovery',
  'token-ledger-maintenance',
  'watchlist-check',
])

export function defaultSystemJobs(now: number): SchedulerJob[] {
  return [
    systemJob('memory-maintenance', 'Memory maintenance', new SchedulerSchedule({ kind: 'cron', expr: '17 3 * * *', tz: 'Asia/Shanghai' }), now),
    systemJob('runtime-maintenance', 'Runtime event maintenance', new SchedulerSchedule({ kind: 'cron', expr: '37 3 * * *', tz: 'Asia/Shanghai' }), now),
    systemJob('team-stale-recovery', 'Team stale recovery', new SchedulerSchedule({ kind: 'every', every_ms: 60 * 60 * 1000 }), now),
    systemJob('token-ledger-maintenance', 'Token ledger maintenance', new SchedulerSchedule({ kind: 'cron', expr: '47 3 * * *', tz: 'Asia/Shanghai' }), now),
    systemJob('watchlist-check', 'Watchlist heartbeat', new SchedulerSchedule({ kind: 'every', every_ms: 6 * 60 * 60 * 1000 }), now),
  ]
}

export function isSystemJob(jobId: string): boolean {
  return SYSTEM_JOB_IDS.has(String(jobId || ''))
}

function systemJob(id: string, name: string, schedule: SchedulerSchedule, now: number): SchedulerJob {
  return SchedulerJob.create({
    jobId: id,
    name,
    schedule,
    payload: new SchedulerPayload({ kind: 'system_event', message: id, meta: { system_event: id } }),
    protected: true,
    purpose: `${name} system job`,
    now,
  })
}
